src/Interpreters/GraceHashJoin.h

```diff
@@ -0,0 +1,135 @@
+ *
+ * 1) During the first stage we accumulate blocks of the right table via @addJoinedBlock.
+ * Each input block is split into multiple buckets based on the hash of the row join keys.
+ * The first bucket is added to the in-memory HashJoin, and the remaining buckets are written to disk for further processing.
+ * When the size of HashJoin exceeds the limits, we double the number of buckets.
+ * There can be multiple threads calling addJoinedBlock, just like @ConcurrentHashJoin.
+ *
+ * 2) At the second stage we process left table blocks via @joinBlock.
+ * Again, each input block is split into multiple buckets by hash.
+ * The first bucket is joined in-memory via HashJoin::joinBlock, and the remaining buckets are written to the disk.
+ *
+ * 3) When the last thread reading left table block finishes, the last stage begins.
+ * Each @DelayedJoiningBlocksProcessor calls repeatedly @getDelayedBlocks until there are no more unfinished buckets left.
+ * Inside @getDelayedBlocks we select the next not processed bucket, load right table blocks from disk into in-memory HashJoin,
+ * And then join them with left table blocks.
+ *
+ * After joining the left table blocks, we can load non-joined rows from the right talble for RIGHT/FULL JOINs.
+ * Note that non-joined rows are processed in multiple threads, unlike HashJoin/ConcurrentHashJoin/MergeJoin.
+ */
+class GraceHashJoin final : public IJoin
+{
+    class FileBucket;
+    class DelayedBlocks;
+    class InMemoryJoin;
+
+    using Buckets = std::vector<std::shared_ptr<FileBucket>>;
+    using BucketsSnapshot = std::shared_ptr<const Buckets>;
+    using InMemoryJoinPtr = std::unique_ptr<InMemoryJoin>;
+
+public:
+    GraceHashJoin(
+        ContextPtr context_, std::shared_ptr<TableJoin> table_join_, const Block & right_sample_block_, bool any_take_last_row_ = false);
+
+    ~GraceHashJoin() override;
+
+    const TableJoin & getTableJoin() const override { return *table_join; }
+
+    bool addJoinedBlock(const Block & block, bool check_limits) override;
+    void checkTypesOfKeys(const Block & block) const override;
+    void joinBlock(Block & block, std::shared_ptr<ExtraBlock> & not_processed) override;
+
+    void setTotals(const Block & block) override;
+    const Block & getTotals() const override;
+
+    size_t getTotalRowCount() const override;
+    size_t getTotalByteCount() const override;
+    bool alwaysReturnsEmptySet() const override;
+
+    bool supportParallelJoin() const override { return true; }
+
+    std::shared_ptr<NotJoinedBlocks>
+    getNonJoinedBlocks(const Block & left_sample_block, const Block & result_sample_block, UInt64 max_block_size) const override;
+
+    /// Open iterator over joined blocks.
+    /// Must be called after all @joinBlock calls.
+    std::unique_ptr<IDelayedJoinedBlocksStream> getDelayedBlocks(IDelayedJoinedBlocksStream * prev_cursor) override;
+
+    static bool isSupported(const std::shared_ptr<TableJoin> & table_join, bool for_auto = false);
```

Comment:
Unclear if it's supported for FULL or RIGHT.
`02274_full_sort_join_nodistinct` doen't pass, so probably not yet.